### DISCUSSION TITLE
[@types/node] events: Strong types for EventEmitter events

### DISCUSSION THREAD

**Discussion by eritbh** [DefinitelyTyped/DefinitelyTyped#55298] [Issues with a @types package] [answered]
I'm not sure if this has been discussed anywhere on this repo in the past or not - I tried searching for related issues, but didn't really find anything, so apologies if this has already been discussed. I'm wondering if it would be possible to add an optional type parameter to the Node `EventEmitter` typings to give a standardized way to represent strongly-typed events in Node projects. I'm thinking about making something like this possible out of the box:

```ts
interface MyEvents {
  foo: [string, number];
  bar: [boolean];
}
const myEventEmitter = new EventEmitter<MyEvents>();

myEventEmitter.on('bar', eventValue => {
  // eventValue is a boolean
});

// @ts-expect-error - ['foo', false] not assignable to ['foo', string, number] | ['bar', boolean]
myEventEmitter.emit('foo', false);
```

I've seen similar patterns being used by other projects, using either subclasses of `EventEmitter` or casting the emitter variable to a custom type in order to achieve the type safety desired. I think it'd be good to have this functionality built into `@types/node` to avoid projects having to use their own solutions that introduce quirks and inconsistencies between libraries. For example, here's a partial example of how to achieve the above:

```ts
interface Events<E extends {[eventName in keyof E]: any[]}> {
  on<K extends keyof E>(eventName: K, listener: (...args: E[K]) => void): this;
  emit<K extends keyof E>(eventName: K, ...args: E[K]): boolean;
  // other EventEmitter methods you care about...
}

const myEventEmitter = new EventEmitter() as Events<MyEvents>;
```

One concern I have is that event emitters aren't actually limited to emitting known event names. Projects with strongly-typed events often type their emitters in such a way that using an unrecognized event is disallowed, which prevents typo issues, but others are fine with the user emitting custom events on their listeners. Can this be designed to allow typings authors that choice?

[a comment by an automated account is not shown]

**Comment by SimonSchick** [comment]
There are 2 issues:
- Backwards compability.
- TypeScripts not really supported inheriting overloads (eg. on, off, etc.)

I've tried a couple times to implement generic EventEmitters but it never quite worked out.

**Reply by eritbh** [reply]
@SimonSchick Thanks for your reply. I may be misunderstanding what the issues here are; correct me if I'm wrong about any of this.

Backwards compatibility with old Typescript versions is actually the main reason I haven't tried to PR this myself; I'm not confident with the system for remapping files for different Typescript versions. Inheriting overloads doesn't seem like a problem to me, though; from what I can tell, overloads aren't necessary to get this working, and subclassing EventEmitter seems to work as expected for simple cases at least.

```ts
// types/node/events.d.ts
type EventMap<T> = {[eventName in keyof T]: any[]};
type DefaultEventMap = {[eventName: string | symbol]: any[]};
// ...
namespace NodeJS {
    interface EventEmitter<T extends EventMap<T> = DefaultEventMap> {
        on<K extends keyof T>(eventName: K, listener: (...args: T[K]) => void): this;
        // other declared methods also rewritten using generics like this...
    }
}

// consumerCode.ts
class PlainSubclass extends events.EventEmitter {
    constructor () {
        super();
        this.emit('foo', 42);
        this.on('bar', () => {});
    }
}
class TypedSubclass extends events.EventEmitter<{
    foo: [string, number];
    bar: [boolean];
}> {
    constructor () {
        super();
        this.emit('foo', 'hello', 42);
        this.on('bar', value => {
            value // inferred as boolean
        });
        // @ts-expect-error
        this.emit('bar', 'hello');
    }
}
```

Is there some case this solution is inadequate for that I'm missing, or is the main bottleneck just the need for TS 4.4 features to get it working?

**Reply by malthe** [reply]
@SimonSchick how do we get the ball rolling on this? Events are everywhere, the lack of generics here are causing fragmentation in the ecosystem.

With the typings suggested above by @eritbh, what are some remaining issues – if any?

**Reply by malthe** [reply, marked as answer]
See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68313.
